File: apps/ResidentApp.Android/app/src/main/java/com/acls/resident/ui/screens/register/RegisterViewModel.kt
package com.acls.resident.ui.screens.register

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.acls.resident.data.remote.toApiException
import com.acls.resident.domain.repository.AuthRepository
import com.acls.resident.ui.ErrorCodeMapper
import com.google.gson.Gson
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import retrofit2.HttpException
import javax.inject.Inject

sealed class RegisterUiState {
    data object Idle : RegisterUiState()
    data object Loading : RegisterUiState()
    data object Success : RegisterUiState()
    data class Error(val message: String) : RegisterUiState()
}

@HiltViewModel
class RegisterViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val gson: Gson,
    @ApplicationContext private val appContext: Context
) : ViewModel() {

    private val _uiState = MutableStateFlow<RegisterUiState>(RegisterUiState.Idle)
    val uiState: StateFlow<RegisterUiState> = _uiState.asStateFlow()

    fun register(
        invitationToken: String,
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        phone: String?
    ) {
        viewModelScope.launch {
            _uiState.value = RegisterUiState.Loading
            _uiState.value = try {
                authRepository.register(invitationToken, email, password, firstName, lastName, phone)
                RegisterUiState.Success
            } catch (e: HttpException) {
                RegisterUiState.Error(ErrorCodeMapper.getMessage(appContext, e.toApiException(gson)))
            } catch (e: Exception) {
                RegisterUiState.Error(ErrorCodeMapper.getMessage(appContext, "System.InternalError"))
            }
        }
    }

    fun resetState() {
        _uiState.value = RegisterUiState.Idle
    }
}

File: apps/ResidentApp.iOS/Sources/RegisterViewModel.swift
import Foundation

// MARK: - RegisterViewModel

@MainActor
final class RegisterViewModel: ObservableObject {
    @Published private(set) var isLoading = false
    @Published private(set) var errorMessage: String?

    private let repository: AuthRepository

    init(repository: AuthRepository = AuthRepository()) {
        self.repository = repository
    }

    func register(
        invitationToken: String,
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        phone: String,
        tokenStore: TokenStore
    ) async {
        guard !firstName.isEmpty, !lastName.isEmpty, !email.isEmpty, !password.isEmpty else {
            errorMessage = NSLocalizedString("error_validation_failed", comment: "")
            return
        }
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            let response = try await repository.register(
                invitationToken: invitationToken,
                email: email,
                password: password,
                firstName: firstName,
                lastName: lastName,
                phone: phone.isEmpty ? nil : phone
            )
            tokenStore.save(response: response)
        } catch let error as APIError {
            errorMessage = ErrorCodeMapper.message(for: error)
        } catch {
            errorMessage = ErrorCodeMapper.message(for: .networkError(error))
        }
    }
}
